### PR TITLE
[Merged by Bors] - chore: forward-port leanprover-community/mathlib#18341 and leanprover-community/mathlib#18950

### PR DESCRIPTION
**Pull request by Ruben-VandeVelde**
This also brings the proof of `partialProd_right_inv` in line with mathlib3.

---

- [ ] depends on: https://github.com/leanprover-community/mathlib/pull/18950

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by 101damnations** [comment]
Thank you very much for doing this!

File: Mathlib/Data/Fin/Basic.lean [hunk 1]
```diff
@@ -2504,6 +2504,10 @@ theorem coe_ofNat_eq_mod (m n : ℕ) [NeZero m] :
   rfl
 #align fin.coe_of_nat_eq_mod Fin.coe_ofNat_eq_mod
 
+lemma coe_add_one_eq_cast {n i : ℕ} (h : i < n) : (i + 1 : Fin (n + 1)) = Fin.succ ⟨i, h⟩ := by
+  rw [succ_mk, Fin.eq_mk_iff_val_eq, ←Nat.cast_add_one, coe_ofNat_eq_mod, Nat.mod_succ_eq_iff_lt]
+  exact Nat.succ_lt_succ h
```
Comment (eric-wieser) [inline Mathlib/Data/Fin/Basic.lean:2509]: Is this lemma new? It seems rather strange to me. Name-wise, I'd expect
```suggestion
lemma natCast_add_one_eq_succ_mk {n i : ℕ} (h : i < n) : (i + 1 : Fin (n + 1)) = Fin.succ ⟨i, h⟩ := by
  rw [succ_mk, Fin.eq_mk_iff_val_eq, ←Nat.cast_add_one, coe_ofNat_eq_mod, Nat.mod_succ_eq_iff_lt]
  exact Nat.succ_lt_succ h
```

File: Mathlib/Algebra/BigOperators/Fin.lean [hunk 2]
```diff
@@ -272,32 +270,70 @@ theorem partialProd_left_inv {G : Type _} [Group G] (f : Fin (n + 1) → G) :
...
+  have := partialProd_right_inv (1 : G) g
+  simp only [one_smul, coe_eq_castSucc] at this
+  rcases lt_trichotomy (k : ℕ) j with (h | h | h)
+  · rwa [succAbove_below, succAbove_below, this, contractNth_apply_of_lt]
+    · assumption
+    · rw [castSucc_lt_iff_succ_le, succ_le_succ_iff, le_iff_val_le_val]
+      exact le_of_lt h
+  · rw [succAbove_below, succAbove_above, partialProd_succ, castSucc_fin_succ, ← mul_assoc,
+      this, contractNth_apply_of_eq]
+    · assumption
```
Comment (eric-wieser) [inline Mathlib/Algebra/BigOperators/Fin.lean:325]: I think this could do with a porting note: why doesn't `rwa` work here?

**Comment by eric-wieser** [comment]
> This also brings the statement of `partialProd_right_inv` in line with mathlib3.

I've opened leanprover-community/mathlib#18950 to make this proof easier to port faithfully, without needing any new lemmas.

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib#18950
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib#18950~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by eric-wieser** [comment]
bors d+

Thanks!

**Pull request closed without merging**